Change PR06 logic to only fail when type is used standalone

amoeba opened the pull request:
Closes https://github.com/numpy/numpydoc/issues/446

This is a draft PR for addressing https://github.com/numpy/numpydoc/issues/446 which hasn't yet been discussed. For reviewing, I'd pay special attention to how I went the logic in `validate.py`, especially the regex.

rossbar commented on numpydoc/validate.py:
```diff
@@ -584,7 +584,15 @@ def validate(obj_name):
                     ("string", "str"),
                 ]
                 for wrong_type, right_type in common_type_errors:
-                    if wrong_type in doc.parameter_type(param):
```
Can't we just change the `in` to `==` instead?

amoeba replied:
Almost, but this needs to support more complex type descriptions than "str" such as "list of str". The regex could be improved to avoid the nested ifs. Let me know if you want me to give that a shot.

rossbar replied:
Good point - then how about `if wrong_type in set(doc.parameter_type(param).split()):`? I tend toward trying to keep the logic as simple and short as possible, but if a regex is necessary then so be it!

amoeba replied:
That was close and got me to where it needed to go. Your above doesn't handle punctuation (specifically `,`) so I changed it to use `re.split` to split on `\W`. That worked nicely.

amoeba commented:
Thanks for the review @rossbar. I think I've addressed your review comment via https://github.com/numpy/numpydoc/pull/447/commits/0990db168b26ba6aa29a2e2276a753ad4919c704. This is ready for another look.

rossbar reviewed (approved):
I took the liberty of running black to make sure the formatting was consistent but other than that this LGTM. Thanks @amoeba !

rossbar commented:
The CI failures are unrelated - see pydata/pydata-sphinx-theme#1130.

rossbar commented:
I'm going to go ahead and get this in since CI has been patched. Thanks for raising+fixing @amoeba

amoeba commented:
Thank you @rossbar!
